proto/pulumi/provider.proto

```diff
@@ -89,8 +93,34 @@ service ResourceProvider {
     rpc GetMappings(GetMappingsRequest) returns (GetMappingsResponse) {}
 }
 
+message ParameterizeRequest {
+    message ParametersArgs {
+        repeated string args = 1;
+    }
+
+    message ParametersValue {
+        // The key (a valid package name) for this sub-schema parameterization.
+        string key = 1;
+        // The package version this was sent from.
+        string version = 2;
+        // The embedded value from the package.
+        google.protobuf.Value value = 3;
+    }
+
+    oneof parameters {
+        ParametersArgs args = 1; // arguments from the command line.
+        ParametersValue value = 2; // values from a generated package.
+    }
+}
+
+message ParameterizeResponse {
+    // The key (a valid package name) for this parameterization.
+    string key = 1;
+}
```

Comment:
Looking at this again, it feels like we should return at least a key & version here. A `key` is not enough to uniquely identify a parameterization. 

The alternative is to bake into all providers logic that says "if I am parameterized twice with two providers that have the same key, return an error". I don't think this will be an uncommon scenario, say when we upgrade the version of a provider and also destroy some resources (on the old provider version).

Reply:
Yeh that's reasonable, we'd need to add that to GetSchema as well.

Reply:
> The alternative is to bake into all providers logic that says "if I am parameterized twice with two providers that have the same key, return an error".

I'm currently leaning a bit towards process isolation and thus actually only ever parameterising a provider process once. This is simple for codegen (we only ever codegen one thing at a time anyway), and it makes Configure and versioning easier to handle during deployment time. Only awkward part is where it leaves debug attach, but that already has issues with explicit providers and multiple versions so I think we just need an overhaul there.

Reply:
I think process isolation would be simpler and easier to make "correct". If we are running into performance constraints we can relax this requirement later.